gateway/load_balancer.go

```diff
@@ -0,0 +1,88 @@
+package gateway
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+)
+
+type roundRobinLoadBalancer struct {
+	hostname     string
+	ips          []string
+	currentIndex int
+	transport    http.RoundTripper
+	sync.RWMutex
+}
+
+func newRoundRobinLoadBalancer(hostname string) *roundRobinLoadBalancer {
+	lb := &roundRobinLoadBalancer{
+		hostname:  hostname,
+		transport: http.DefaultTransport,
+	}
+
+	// Resolve IPs initially
+	ips, err := net.LookupIP(lb.hostname)
+	if err != nil {
+		log.Printf("Failed to resolve IPs for hostname %s: %v", lb.hostname, err)
+	} else {
+		lb.ips = ipsToStrings(ips)
+	}
+
+	return lb
+}
+
+func (lb *roundRobinLoadBalancer) roundTrip(req *http.Request) (*http.Response, error) {
+	lb.RLock()
+	defer lb.RUnlock()
+
+	if len(lb.ips) == 0 {
+		// TODO: replace format error with a log statement
+		return nil, fmt.Errorf("no IP addresses available")
+	}
+
+	ip := lb.ips[lb.currentIndex%len(lb.ips)]
+	req.URL.Host = strings.Replace(req.URL.Host, lb.hostname, ip, 1)
+	lb.currentIndex++
+
+	return lb.transport.RoundTrip(req)
+}
+
+// Refresh IPs periodically
+func (lb *roundRobinLoadBalancer) refreshIPs(refreshInterval time.Duration) {
+	for {
+		ips, err := net.LookupIP(lb.hostname)
+		if err != nil {
+			// TODO: replace std library log package with logrus
+			log.Printf("Failed to resolve IPs for hostname %s: %v", lb.hostname, err)
+		} else {
+			lb.Lock()
+			lb.ips = ipsToStrings(ips)
+			lb.currentIndex = 0
```

Comment:
I am not sure about this, as resetting to zero means we will use the first resolved IP more than others.

But I think tests will give the final veredict: Include some tests that prove every ip address gets similar number of requests over many refresh dns.